fix: show all messages associated with connection failures

danielfrankcom opened the pull request:
This PR fixes #1069.

It updates the connection failure logic to preserve all error messages for the various attempts. Previously, only the message from the final attempt was shown the user, which created confusing situations as described in #1069.

Below is an example of the new message format:

```
psycopg.OperationalError: connection failed: connection to server at "1.2.3.4", port 5432 failed: FATAL:  unable to accept connection, access denied
DETAIL:  Session Id: ghijklmnp4yquucuzkab7hhka
Multiple connection attempts failed. All failures were:
- host: 'redacted.for.github', port: None, hostaddr: '2200:1f18:1f72:a404:61e8:e38a:6be2:fake': connection failed: connection to server at "2200:1f18:1f72:a404:61e8:e38a:6be2:fake", port 5432 failed: FATAL:  unable to accept connection, access denied
DETAIL:  Session Id: abcdefloavxpiw6oweks7x7fcu
- host: 'redacted.for.github', port: None, hostaddr: '1.2.3.4': connection failed: connection to server at "1.2.3.4", port 5432 failed: FATAL:  unable to accept connection, access denied
DETAIL:  Session Id: ghijklmnp4yquucuzkab7hhka
```

I made a few decisions which could be discussed:
- I decided to keep the `-` pushed to the left rather than indent it, as some messages contain multiple lines. We shouldn't need to worry about artificially indenting multi-line messages, and this is made more complicated by the fact that on my terminal the first line of these messages is long enough to wrap itself. I think it's easier to see the delineation if the `-` is pushed up against the left rather than indented in these cases of multi-line or wrapping messages.
- I replaced the debug message and included the connection information in the message itself. As a user that was trying to use these messages to diagnose a problem in #1069, I think having that information is super valuable for diagnostic purposes. Though the example I posted above includes duplicate information, not all messages do and it depends on the type of failure. A timeout exception for example doesn't show any information about which host/port it was trying to connect to. If you end up with different exceptions for different hosts, it's extremely valuable to know which host produced which error.

dvarrazzo commented on psycopg/psycopg/errors.py:
```diff
@@ -1757,3 +1757,4 @@ class IndexCorrupted(InternalError,
 # instance in a certain `Cursor.execute()`.
 
 _NO_TRACEBACK = (Error, KeyboardInterrupt, CancelledError)
+_NoTracebackException = Union[Error, KeyboardInterrupt, CancelledError]
```
Unrelated, but I just noticed (https://github.com/psycopg/psycopg/issues/1055#issuecomment-2859787149) that it's wrong here to trap KeyboardInterruplt: instead of stopping we just move to the following attempt.

So the except should probably just trap `psycopg.Error`, not the other we trap in queries.

dvarrazzo replied:
This was addressed by #1077.

dvarrazzo commented on psycopg/psycopg/connection.py:
```diff
@@ -95,27 +95,44 @@ def connect(
         timeout = timeout_from_conninfo(params)
         rv = None
         attempts = conninfo_attempts(params)
+        connection_errors: list[tuple[e._NoTracebackException, str]] = []
```
You could use `list[tuple[BaseException, str]]` or even `list[tuple[Exception, str]]` if we stop catching `KeyboardInterrupt/CancelledError` here.

danielfrankcom replied:
Agreed, removing those other exceptions would make the types a lot simpler here

danielfrankcom replied:
Should be addressed in [982dda4](https://github.com/psycopg/psycopg/commit/982dda4f5c99832d95e0e447ba6c16eb9122a24a). I've left the `_NO_TRACEBACK` variable since it's used in quite a few places and may still be needed in the future.

dvarrazzo commented on psycopg/psycopg/connection.py:
```diff
@@ -95,27 +95,44 @@ def connect(
+                attempt_details = "host: {}, port: {}, hostaddr: {}".format(
+                    repr(attempt.get("host")),
+                    repr(attempt.get("port")),
+                    repr(attempt.get("hostaddr")),
```
Is this the most useful order in which to show these 3 details or you would rather see them in a different error to spot problems more easily (hostaddr first maybe?)

danielfrankcom replied:
I'm not sure - my thought was there may be multiple hosts, and then multiple addresses associated with each host. I tried to do it in the order of what I thought was likely to have the least variations first, moving to the most variations on the right.

dvarrazzo replied:
Ok, let's leave it this way.

dvarrazzo commented on psycopg/psycopg/connection.py:
```diff
@@ -95,27 +95,44 @@ def connect(
-            assert last_ex
-            raise last_ex.with_traceback(None)
+            num_errors = len(connection_errors)
+            assert num_errors > 0
```
I don't think this assert is useful. `assert last_ex` was necessary to please MyPy, but this list will never be empty, and even if it was, IndexError or AssertError don't make any difference.

danielfrankcom replied:
Agreed, I can remove this.

danielfrankcom replied:
Addressed in [982dda4](https://github.com/psycopg/psycopg/commit/982dda4f5c99832d95e0e447ba6c16eb9122a24a)

dvarrazzo commented on psycopg/psycopg/connection.py:
```diff
@@ -95,27 +95,44 @@ def connect(
+                message_lines.extend(formatted_attempts)
+                enhanced_message = "\n".join(message_lines)
+                enhanced_exception = last_exception_type(enhanced_message)
+                raise enhanced_exception.with_traceback(None) from last_exception
```
Does it make sense to raise `from last_exception`? I don't think it adds much value. Maybe a simple exception is better.

danielfrankcom replied:
I think this is a hold-over from before you mentioned they were already flattened, so I was trying to preserve the origin of the exception. Since they're flat, I don't think this makes sense. Will remove.

danielfrankcom replied:
Addressed in [982dda4](https://github.com/psycopg/psycopg/commit/982dda4f5c99832d95e0e447ba6c16eb9122a24a)

dvarrazzo commented on tests/test_connection.py:
```diff
@@ -37,6 +37,46 @@ def test_connect_bad(conn_cls):
+
+    msg = str(e)
+    assert 'database "nosuchdb" does not exist' in msg
+    assert "Multiple connection attempts failed" not in msg
```
Please put this string in a const and use the same const for both the tests, otherwise, if we change the string this test would become invalid (it would pass even if we break something).

danielfrankcom replied:
Addressed in [982dda4](https://github.com/psycopg/psycopg/commit/982dda4f5c99832d95e0e447ba6c16eb9122a24a)

dvarrazzo commented on psycopg/psycopg/connection_async.py:
```diff
@@ -111,27 +111,46 @@ async def connect(
+            else:
+                formatted_attempts = []
+                for error, attempt_details in connection_errors:
+                    formatted_attempts.append(f"- {attempt_details}: {str(error)}")
```
`str()` is implicit
```suggestion
                    formatted_attempts.append(f"- {attempt_details}: {error}")
```

danielfrankcom replied:
Addressed in [982dda4](https://github.com/psycopg/psycopg/commit/982dda4f5c99832d95e0e447ba6c16eb9122a24a)

danielfrankcom commented:
Recent pushes are attempting to fix the tests for `crdb`. Haven't taken a look through comments yet.

danielfrankcom commented:
In the latest push I modified the first unit test to depend on a timeout error instead of the invalid DB error. It seems CRDB doesn't produce an error when the DB doesn't exist, so I need to reply on a different type of failure for that case.

This was a bit confusing as the original test directly above the 2 I added gets an `OperationalError` from CRDB, but it seems like it's because the DSN isn't configured. If I set the `dsn` and override the `dbname`, the CRDB connection still succeeds which isn't what I want for that unit test.

dvarrazzo commented:
I am making the change in #1077 where we log and continue attempts only when a database error is returned, and we exit immediately if a ctrl-c is received. I would say let's first merge that MR, then we can rebase this on top, so the type annotation can be simplified.

dvarrazzo commented:
> Recent pushes are attempting to fix the tests for `crdb`. Haven't taken a look through comments yet.

We can just skip the test on CRDB. This is a purely client behaviour, different vendors will not make a relevant difference.

danielfrankcom commented:
> > Recent pushes are attempting to fix the tests for `crdb`. Haven't taken a look through comments yet.
> 
> We can just skip the test on CRDB. This is a purely client behaviour, different vendors will not make a relevant difference.

We can do that if you would prefer, but there should be no need with the latest PR version. I switched the forced exception type to a timeout, which worked and passed all tests including on CRDB.

Regarding #1077 and rebasing this one afterwards - that sounds good to me.

dvarrazzo commented on tests/test_connection.py:
```diff
@@ -37,6 +39,51 @@ def test_connect_bad(conn_cls):
+    assert MULTI_FAILURE_MESSAGE not in msg
+
+
+def test_connect_error_multi_hosts_each_message_preserved(conn_cls, monkeypatch):
+    # This is unreasonably short as it is expected to fail as fast as possible.
+    monkeypatch.setattr(psycopg.conninfo, "_DEFAULT_CONNECT_TIMEOUT", 0.1)
+
+    # IPv4 address blocks reserved for documentation.
+    # https://datatracker.ietf.org/doc/rfc5737/
+    args = {"host": "192.0.2.1,198.51.100.1", "port": "1234,5678"}
```
This is cool, I didn't know about these unroutable addresses! Thank you for the reference.

So far, in order to test with a host that doesn't reply, we opened a listening socket with [proxy.deaf_port](https://github.com/psycopg/psycopg/blob/24ad0eb967188702b6fccabb2d1b46a7f6a8b7c7/tests/fix_proxy.py#L105-L118) but if these IPs work reliably we can probably do without.

You can set `connect_timeout` in the connection string instead of monkeypatching. However, 0.1s is a slow test :slightly_smiling_face: so we need to mark it so.

```suggestion
@pytest.mark.slow
def test_connect_error_multi_hosts_each_message_preserved(conn_cls):
    # IPv4 address blocks reserved for documentation.
    # https://datatracker.ietf.org/doc/rfc5737/
    args = {"host": "192.0.2.1,198.51.100.1", "port": "1234,5678", "connect_timeout": 0.1}
```

same for both tests, both sync and async.

danielfrankcom replied:
I didn't use the proxy in this case since I needed 2 hosts and 2 ports that would reliably fail and it seemed like the proxy only exposed 1 of each. If you have a suggestion for how to do that with the proxy then I'm happy to swap, but the documentation IPs should work reliably and have done so in my testing.

I'll remove the monkeypatch and mark the tests accordingly, thanks for the tips on this.

dvarrazzo replied:
These tests hosts are ok. In other tests we have in the test suite probably we monkeypatch the dns to inject expected failures, but I don't think it's needed here.

danielfrankcom replied:
Should be addressed by [24c2995](https://github.com/psycopg/psycopg/commit/24c299561e3f13793029fa98706aaac637fce92d)

danielfrankcom commented on psycopg/psycopg/errors.py:
```diff
@@ -1756,4 +1755,4 @@ class IndexCorrupted(InternalError,
 # instance in a certain `Cursor.execute()`.
 
-_NO_TRACEBACK = (Error, KeyboardInterrupt, CancelledError)
+_NO_TRACEBACK = (Error,)
```
Ah I see, I misunderstood our previous discussion. I'll rebase onto the other change now since that should remove any need for me to worry about this.

danielfrankcom commented:
Latest push rebases onto #1077. I will push again shortly after addressing the test comments.

danielfrankcom commented:
Latest push addresses unit test comments. I removed the monkeypatching, and used the deaf proxy for the first test since it only required 1 host/port. Both tests were updated to use the `connect_timeout` parameter, though I had to set them to a 2 second timeout due to the logic [here](https://github.com/psycopg/psycopg/blob/3.2.7/psycopg/psycopg/conninfo.py#L142) which the monkeypatch was bypassing.

dvarrazzo commented:
Thank you very much, merged!

danielfrankcom commented:
Excellent, thank you
